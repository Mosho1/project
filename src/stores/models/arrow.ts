import { types } from 'mobx-state-tree'
import { Socket } from './socket';
import { calculateBezierPoints } from '../utils/utils';
import { pouch } from '../utils/pouchdb-model';



export const Arrow = pouch.model('Arrow', {
    input: types.reference(Socket),
    output: types.reference(Socket)
}).views(self => {
    return {
        get points() {
            return calculateBezierPoints(
                self.output.x,
                self.output.y,
                self.input.x,
                self.input.y,
            );
        },
        get isExec() {
            return self.input.isExec;
        }
    };
});
