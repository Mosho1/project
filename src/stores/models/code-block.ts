import { types } from 'mobx-state-tree';
import { optionalIdentifierType } from '../utils/utils';

const parseFunction = (value: string) => {
    const fn = eval(`(${value})`);
    if (typeof fn !== 'function') throw new Error(`${value} is not a valid function`);
    return fn;
};

export const codeType = types.custom<string, Function>({
    name: 'CodeBlock',
    fromSnapshot(value: string) {
        return parseFunction(value);
    },
    toSnapshot(value: Function) {
        return value.toString();
    },
    getValidationMessage(value: string) {
        try {
            parseFunction(value);
            return '';
        } catch (e) {
            return `value "${value}" is Not a valid function ${e}`;
        }
    },
    isTargetType(value: any) {
        return value instanceof Function;
    }
});

export const editableTypes = types.enumeration('editableTypes', [
    'string',
    'number'
]);

export type IEditableTypes = typeof editableTypes.Type;

const nonEditableTypes = types.enumeration('nonEditableTypes', [
    'any',
    'void'
]);

export const typeNames = ['string', 'number', 'any', 'void'];

export const primitiveTypes = types.union(editableTypes, nonEditableTypes);

export const CodeBlockIO = types.model('CodeBlockIO', {
    id: optionalIdentifierType,
    name: types.optional(types.string, ''),
    type: types.optional(primitiveTypes, 'any'),
    defaultValue: types.maybe(types.string)
});

export const CodeBlock = types.model('CodeBlock', {
    id: optionalIdentifierType,
    name: types.string,
    code: codeType,
    runOnStart: types.optional(types.boolean, false),
    values: types.optional(types.array(CodeBlockIO), []),
    inputs: types.optional(types.array(CodeBlockIO), []),
    returns: types.maybe(types.refinement(CodeBlockIO, io => io.type !== 'any') as typeof CodeBlockIO),
    execInputs: types.optional(types.array(CodeBlockIO), []),
    execOutputs: types.optional(types.array(CodeBlockIO), []),
});

type ICodeBlockType = typeof CodeBlock.Type;

type ICodeBlockSnapshotType = typeof CodeBlock.SnapshotType;
export interface ICodeBlock extends ICodeBlockType { };
export interface ICodeBlockSnapshot extends ICodeBlockSnapshotType { };

type ICodeBlockIOType = typeof CodeBlockIO.Type;

type ICodeBlockIOSnapshotType = typeof CodeBlockIO.SnapshotType;
export interface ICodeBlockIO extends ICodeBlockIOType { };
export interface ICodeBlockIOSnapshot extends ICodeBlockIOSnapshotType { };