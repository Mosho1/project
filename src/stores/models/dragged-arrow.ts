import { types } from 'mobx-state-tree';
import { calculateBezierPoints } from '../utils/utils';

export const DraggedArrow = types.model('DraggedArrow', {
    startX: types.number,
    startY: types.number,
    endX: types.number,
    endY: types.number,
    fromInput: types.boolean,
}).views(self => {
    return {
        get points() {
            const points = [self.startX, self.startY];
            if (self.fromInput) {
                points.unshift(self.endX, self.endY);
            } else {
                points.push(self.endX, self.endY);
            }
            return calculateBezierPoints(...points);
        }
    };
}).actions(self => {
    const start = (x: number, y: number) => {
        self.startX += x;
        self.startY += y;
        return self;
    };

    const end = (x: number, y: number) => {
        self.endX += x;
        self.endY += y;
        return self;
    };

    return {
        start, end
    };
});

type IDraggedArrowType = typeof DraggedArrow.Type;
export interface IDraggedArrow extends IDraggedArrowType { };