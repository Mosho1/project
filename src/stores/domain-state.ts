import {
    types,
    getEnv,
    applySnapshot,
    getSnapshot,
    detach,
    hasParent,
    getParent
} from 'mobx-state-tree'
import { pouch } from './utils/pouchdb-model';
import { models, modelTypes } from './models/index';
import { ContextMenu } from './context-menu';
import { ICodeBlock, ICodeBlockIO } from './models/code-block';
import * as codeBlocks from './functions';
import { values } from './utils/utils';
import { run, stop } from './run';
import { SocketTypeEnum, areSocketsCompatible } from './models/socket';
import { observable } from 'mobx';

const Position = types.model('Position', {
    x: types.number,
    y: types.number
});

const Stage = types.model('Stage', {
    scale: types.optional(types.number, 1),
    position: types.optional(Position, { x: 0, y: 0 })
}).views(self => ({
    get store(): null | IStore {
        if (!hasParent(self, 1)) return null;
        return getParent(self, 1);
    },
})).actions(self => ({
    move(dx: number, dy: number) {
        self.position.x += dx;
        self.position.y += dy;
        return self;
    },
    setScale(scale: number) {
        self.scale = scale;
    },
    handleScale(e: WheelEvent) {
        const mousePointTo = {
            x: e.clientX / self.scale - self.position.x / self.scale,
            y: e.clientY / self.scale - self.position.y / self.scale,
        };

        const scaleBy = e.ctrlKey ? 1.3 : 1.1;

        const newScale = e.deltaY < 0 ? self.scale * scaleBy : self.scale / scaleBy;

        if (newScale > 3 || newScale < 0.4) return;

        self.scale = newScale;

        self.position = {
            x: -(mousePointTo.x - e.clientX / newScale) * newScale,
            y: -(mousePointTo.y - e.clientY / newScale) * newScale
        }
    },
    getAbsolutePosition(x: number, y: number) {
        return {
            x: x - self.position.x,
            y: y - self.position.y,
        }
    }
}));

type IStageType = typeof Stage.Type;
export interface IStage extends IStageType { };

export const Store = pouch.store('Store', {
    boxes: types.optional(types.map(models.Box), {}),
    sockets: types.optional(types.map(models.Socket), {}),
    arrows: types.optional(types.map(models.Arrow), {}),
    codeBlocks: types.optional(types.map(models.CodeBlock), {}),
    selection: types.optional(types.array(types.reference<modelTypes['Box']>(models.Box)), []),
    draggedArrow: types.maybe(models.DraggedArrow),
    draggedRect: types.maybe(models.DraggedRect),
    draggedFromSocket: types.maybe(types.reference<modelTypes['Socket']>(models.Socket)),
    contextMenu: types.maybe(ContextMenu),
    stage: types.optional(Stage, {}),
}).volatile(_ => ({
    running: false,
    breakpointCallback: null as Function | null,
    breakPosition: null as modelTypes['Box'] | null
}))
    .actions(self => ({
        setDraggedFromSocket(socket: modelTypes['Socket'] | null) {
            self.draggedFromSocket = socket;
        },
        setBreakpoint(box: modelTypes['Box'], cb: Function) {
            self.breakpointCallback = cb;
            self.breakPosition = box;
        },
        continueAfterBreakpoint() {
            self.breakpointCallback = null;
            self.breakPosition = null;
        }

    }))
    .actions(self => {
        const addBox = (name: string, x: number, y: number, code: ICodeBlock) => {
            const { inputs, returns, execInputs, execOutputs, values } = code;
            const boxValues = values.map(v => ({ name: v.name, value: v.defaultValue || '' }));
            const box = models.Box.create({ name, code, values: boxValues, ...self.stage.getAbsolutePosition(x - 75, y - 60) });
            for (const input of inputs) {
                addSocketToBox(box, 'input', input);
            }
            if (returns && returns.type !== 'void') {
                addSocketToBox(box, 'output', returns);
            }
            for (const input of execInputs) {
                addSocketToBox(box, 'execInput', input);
            }
            for (const output of execOutputs) {
                addSocketToBox(box, 'execOutput', output);
            }
            self.boxes.put(box);
            return box;
        };
        const addBoxAndArrowIfDragged = (name: string, x: number, y: number, code: ICodeBlock) => {
            const b = addBox(name, x, y, code);
            if (self.draggedFromSocket) {
                for (const socket of b.sockets) {
                    if (areSocketsCompatible(socket, self.draggedFromSocket)) {
                        const a = addArrowFromDraggedSocket(socket);
                        if (a !== null) return true;
                    }
                }
            }
            return false;
        };
        const setSelection = (selection: modelTypes['Box'][]) => {
            self.selection.replace(selection);
            return self;
        };
        const addToSelection = (selection: modelTypes['Box'][]) => {
            self.selection.push(...selection);
            return self;
        };
        const deleteBox = (box: modelTypes['Box']) => {
            self.selection.remove(box);
            for (const socket of box.sockets) {
                deleteArrowsForSocket(socket);
            }

            for (const socket of box.sockets) {
                detach(socket);
            }

            detach(box);
        };
        const deleteSelection = () => {
            const boxes = self.selection.slice(0);
            for (const box of boxes) {
                deleteBox(box);
            }
            setSelection([]);
        };
        const createBox = (name: string, x: number, y: number, code: ICodeBlock) => {
            const box = addBox(name, x, y, code);
            // setSelection(box);
            return box;
        };
        const startDragArrow = (socket: modelTypes['Socket']) => {
            const { x, y } = socket;
            if (socket.isInput && socket.arrows.length > 0) return;
            self.draggedArrow = models.DraggedArrow.create({ startX: x, startY: y, endX: x, endY: y });
            self.setDraggedFromSocket(socket);
        };

        const moveDragArrow = (x: number, y: number) => {
            if (self.draggedArrow) {
                if (self.draggedFromSocket!.isInput) {
                    self.draggedArrow.start(x, y);
                } else {
                    self.draggedArrow.end(x, y);
                }
            }
        };
        const endDragArrow = (
            socket?: modelTypes['Socket'] | null,
            clientX = self.draggedArrow!.endX,
            clientY = self.draggedArrow!.endY,
        ) => {
            self.draggedArrow = null;
            if (!self.draggedFromSocket) return;
            if (socket) {
                addArrowFromDraggedSocket(socket);
            } else if (self.contextMenu) {
                const { isInput, isExec, code } = self.draggedFromSocket;
                let typeFilter: null | ((code: ICodeBlock) => boolean) = null;
                if (isExec) {
                    if (isInput) {
                        typeFilter = (c) => c.execOutputs.some(i => i.type === code.type);
                    } else {
                        typeFilter = (c) => c.execInputs.some(i => i.type === code.type);
                    }
                } else {
                    switch (code.type) {
                        case 'number':
                        case 'string':
                            if (isInput) {
                                typeFilter = (c) => c.returns ? c.returns.type === 'any' || c.returns.type === code.type : false;
                            } else {
                                typeFilter = (c) => c.inputs.some(i => i.type === 'any' || i.type === code.type);
                            }
                    }
                }

                setTimeout(() => self.contextMenu!.toggle(true, clientX, clientY, typeFilter));
            }
        };
        const startDragRect = (x: number, y: number) => {
            self.draggedRect = models.DraggedRect.create({ startX: x, startY: y, endX: x, endY: y });
        };

        const moveDragRect = (x: number, y: number) => {
            if (!self.draggedRect) return;
            self.draggedRect.end(x, y);
        };
        const endDragRect = () => {
            if (!self.draggedRect) return;
            const selection = observable<modelTypes['Box']>([]);
            for (const box of values(self.boxes)) {
                if (self.draggedRect.hasIntersection(box)) {
                    selection.push(box);
                }
            }
            self.selection = selection;
            self.draggedRect = null;
        };
        const hasArrow = (from: modelTypes['Socket'], to?: modelTypes['Socket']) => {
            const [input, output] = from.isInput ? [from, to] : [to, from];
            return Boolean(values(self.arrows).find(a => {
                return (!input || a.input === input) &&
                    (!output || a.output === output);
            }));
        };
        const addArrow = (input: modelTypes['Socket'], output: modelTypes['Socket']) => {
            if (input.isCompatibleWith(output) &&
                !hasArrow(input, output)) {
                let arrow: null | modelTypes['Arrow'] = null;
                arrow = models.Arrow.create({ input, output });
                self.arrows.put(arrow);
                return arrow;
            }

            return null;

        };
        const addArrowFromDraggedSocket = (socket: modelTypes['Socket']) => {
            if (!self.draggedFromSocket) return;
            const [input, output] = socket.isInput ? [socket, self.draggedFromSocket] : [self.draggedFromSocket, socket];
            self.setDraggedFromSocket(null);
            return addArrow(input, output);
        };
        const deleteArrowsForSocket = (socket: modelTypes['Socket']) => {
            for (const arrow of values(self.arrows)) {
                if (arrow.input === socket || arrow.output === socket) {
                    detach(arrow);
                    // arrowsToDelete.push(arrow._id);
                }
            }
            // for (const id of arrowsToDelete) {
            //     self.arrows.delete(id);
            // }
        };

        const runCode = () => {
            if (self.running) {
                if (self.breakpointCallback) {
                    self.breakpointCallback();
                    self.continueAfterBreakpoint();
                }
            } else {
                self.running = true;
                const env = getEnv<{ run: typeof run }>(self);
                env.run(self.boxes, (box, resume) =>
                    self.setBreakpoint(box, resume));
            }
        };

        const stopCode = () => {
            self.running = false;
            self.continueAfterBreakpoint();
            return getEnv(self).stop();
        };

        const addSocketToBox = (box: modelTypes['Box'], type: SocketTypeEnum, code: ICodeBlockIO) => {
            const s = models.Socket.create({ socketType: type, name: code.name, code });
            self.sockets.put(s);
            box.addSocket(s);
            return s;
        };
        const moveBoxOrSelection = (box: modelTypes['Box'], dx: number, dy: number) => {
            if (self.selection.length > 0 && self.selection.indexOf(box) > -1) {
                for (const box of self.selection) {
                    box.move(dx, dy);
                }
            } else {
                box.move(dx, dy);
            }
        };

        return {
            addBox,
            addBoxAndArrowIfDragged,
            hasArrow,
            setSelection,
            addToSelection,
            createBox,
            deleteSelection,
            deleteBox,
            addArrow,
            startDragArrow,
            moveDragArrow,
            endDragArrow,
            startDragRect,
            moveDragRect,
            endDragRect,
            deleteArrowsForSocket,
            runCode,
            stopCode,
            addSocketToBox,
            moveBoxOrSelection,
        };
    })



/*
    The store that holds our domain: boxes and arrows
*/

export const defaults: IStoreSnapshot = {
    boxes: {},
    arrows: {},
    contextMenu: {},
    codeBlocks: { ...codeBlocks.functions },
    selection: [],
    draggedArrow: null,
    draggedFromSocket: null
};

/* istanbul ignore next */
export const getStore = (data?: IStoreSnapshot) => {
    return Store.create({ ...defaults, ...data }, { run, stop });
}

export const replaceStore = (newStore: IStore, oldStore: IStore) => {
    applySnapshot(newStore, { ...getSnapshot(oldStore), codeBlocks: { ...codeBlocks.functions } });
};

type IStoreType = typeof Store.Type;
export interface IStore extends IStoreType { };
type IStoreSnapshotType = typeof Store.SnapshotType;
export interface IStoreSnapshot extends IStoreSnapshotType { };
