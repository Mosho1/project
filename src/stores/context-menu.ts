import { types, hasParent, getParent } from 'mobx-state-tree'
import { IStore } from './domain-state';
import { modelTypes } from './models';
import { values, filterBy } from './utils/utils';
import { codeType, ICodeBlock } from './models/code-block';

export const ContextMenu = types.model('ContextMenu', {
    position: types.optional(types.model({ left: types.number, top: types.number }), { left: 0, top: 0 }),
    isOpen: types.optional(types.boolean, false),
    filter: types.optional(types.string, ''),
    typeFilter: types.maybe(codeType)
}).volatile(_self => ({
    ref: null as null | HTMLElement
}))
    .views(self => ({
        get store(): null | IStore {
            if (!hasParent(self, 1)) return null;
            return getParent(self, 1);
        },
    }))
    .views(self => ({
        get sortedCodeBlocks(): modelTypes['CodeBlock'][] {
            return values(self.store!.codeBlocks).slice().sort((a, b) => a.name > b.name ? 1 : -1);
        }
    }))
    .views(self => ({
        get filteredCodeBlocks(): modelTypes['CodeBlock'][] {
            const filtered = filterBy(self.sortedCodeBlocks, self.filter, b => b.name);
            if (!self.typeFilter) return filtered;
            return filtered.filter(self.typeFilter as (codeBock: ICodeBlock) => boolean);
        }
    }))
    .actions(self => ({
        setRef(ref: null | HTMLElement) {
            self.ref = ref;
        },
        setFilter(value: string) {
            self.filter = value;
        },
        handleContextMenu(clientX: number, clientY: number, contextMenuElement = self.ref) {
            if (!contextMenuElement) return;
            self.isOpen = true;

            const clickX = clientX;
            const clickY = clientY;
            const screenW = window.innerWidth;
            const screenH = window.innerHeight;
            const rootW = contextMenuElement.offsetWidth;
            const rootH = contextMenuElement.offsetHeight;

            const right = (screenW - clickX) > rootW;
            const left = !right;
            const top = (screenH - clickY) > rootH;
            const bottom = !top;

            if (right) {
                self.position.left = clickX + 5;
            }

            if (left) {
                self.position.left = clickX - rootW - 5;
            }

            if (top) {
                self.position.top = clickY + 5;
            }

            if (bottom) {
                self.position.top = clickY - rootH - 5;
            }
        }
    })).actions(self => ({
        _toggle(value: boolean, typeFilter?: null | ((codeBock: ICodeBlock) => boolean)) {
            if (typeof value === 'boolean') self.isOpen = value;
            else self.isOpen = !self.isOpen;
            if (typeFilter) {
                self.typeFilter = typeFilter;
            }
            if (!self.isOpen) {
                self.filter = '';
                self.typeFilter = null;
            }
            return self.isOpen;
        },
    })).volatile(self => ({
        toggle(value: boolean, clientX = 0, clientY = 0, typeFilter?: null | ((codeBock: ICodeBlock) => boolean)) {
            if (self._toggle(value, typeFilter)) {
                self.handleContextMenu(clientX, clientY);
            }
        }
    }));

type IContextMenuType = typeof ContextMenu.Type;
export interface IContextMenu extends IContextMenuType { };