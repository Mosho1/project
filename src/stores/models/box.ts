import { types, getParent, hasParent } from 'mobx-state-tree'
import { pouch } from '../utils/pouchdb-model';
import { Socket, SocketTypeEnum } from './socket';
import { IStore } from '../domain-state';
import { CodeBlock } from './code-block';

interface BoxEditableProps {
    x?: number;
    y?: number;
    width?: number;
    height?: number;
    scaleX?: number;
    scaleY?: number;
}

export const Box = pouch.model('Box',
    {
        name: 'hal',
        x: 0,
        y: 0,
        width: 150,
        value: types.maybe(types.string),
        sockets: types.optional(types.array(Socket), []),
        code: types.reference(CodeBlock)
    })
    .views(self => ({
        get store(): IStore | null {
            if (!hasParent(self)) return null;
            return getParent(self, 2);
        },
        get inputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'input');
        },
        get outputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'output');
        },
        get execInputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'execInput');
        },
        get execOutputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'execOutput');
        },

    }))
    .views(self => ({
        get isSelected() {
            return Boolean(self.store && self.store.selection === self);
        },
        get height() {
            return 50 + Math.max(
                self.execInputs.length + self.inputs.length,
                self.execOutputs.length + self.outputs.length)
                * 30;
        }
    }))
    .actions(self => ({
        move(dx: number, dy: number) {
            self.x += dx;
            self.y += dy;
            return self;
        },
        setProps(props: BoxEditableProps) {
            /* istanbul ignore next */
            Object.assign(self, props);
        },
        addSocket(type: SocketTypeEnum, name = '') {
            const socket = Socket.create({ name: name, socketType: type });
            self.sockets.push(socket);
            return socket;
        },
        setValue(value: string) {
            self.value = value;
        }
    }));

type IBoxType = typeof Box.Type;
export interface IBox extends IBoxType { };
