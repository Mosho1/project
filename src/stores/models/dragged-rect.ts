import { types, hasParent, getParent } from 'mobx-state-tree';
import { IBox } from './box';
import { IStore, IStage } from '../domain-state';

export const DraggedRect = types.model('DraggedRect', {
    startX: types.number,
    startY: types.number,
    endX: types.number,
    endY: types.number,
}).views(self => ({
    get store(): null | IStore {
        if (!hasParent(self, 1)) return null;
        return getParent(self, 1);
    },
    get width() {
        return self.endX - self.startX;
    },
    get height() {
        return self.endY - self.startY;
    },
})).views(self => ({
    get stage(): null | IStage {
        if (!self.store) return null;
        return self.store.stage;
    }
})).views(self => ({
    get absoluteCoords(): { x: number, y: number } {
        if (!self.stage) return { x: self.startX, y: self.startY };
        return self.stage.getAbsolutePosition(
            self.startX,
            self.startY
        );
    }
})).views(self => ({
    get x() {
        return self.width > 0 ? self.absoluteCoords.x : self.absoluteCoords.x + self.width;
    },
    get y() {
        return self.height > 0 ? self.absoluteCoords.y : self.absoluteCoords.y + self.height;
    },
})).views(self => ({
    hasIntersection({ x, y, width, height }: IBox) {
        return !(self.x > x + width ||
            self.x + Math.abs(self.width) < x ||
            self.y > y + height ||
            self.y + Math.abs(self.height) < y);
    }
})).actions(self => {
    const start = (x: number, y: number) => {
        self.startX += x;
        self.startY += y;
        return self;
    };

    const end = (x: number, y: number) => {
        self.endX += x;
        self.endY += y;
        return self;
    };

    return {
        start, end
    };
});

type IDraggedRectType = typeof DraggedRect.Type;
export interface IDraggedRect extends IDraggedRectType { };