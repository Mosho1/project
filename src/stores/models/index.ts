import { Box, IBox } from './box';
import { Socket, ISocket } from './socket';
import { Arrow, IArrow } from './arrow';
import { DraggedArrow, IDraggedArrow } from './dragged-arrow';
import { CodeBlock, ICodeBlock, ICodeBlockIO, CodeBlockIO } from './code-block';

export const models = {
    Box: Box,
    Socket,
    Arrow,
    DraggedArrow,
    CodeBlock,
    CodeBlockIO,
};

export type modelTypes = {
    Box: IBox,
    Socket: ISocket,
    Arrow: IArrow,
    DraggedArrow: IDraggedArrow,
    CodeBlock: ICodeBlock,
    CodeBlockIO: ICodeBlockIO,
};