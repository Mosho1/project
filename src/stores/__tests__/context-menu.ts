import { ContextMenu } from "../context-menu";
import { Store } from '../domain-state';
import { createTestCodeBlock } from '../models/__tests__/code-block';
import { ICodeBlock } from '../models/code-block';

test('sortedCodeBlocks', () => {
    let codeBlocks: { [index: string]: ICodeBlock } = {};
    for (const name of ['zzz', 'ac', 'asef', 'ba', 'esv', 'cef', 'cef', 'ce3', 'dd', 'd']) {
        codeBlocks[name] = createTestCodeBlock({
            id: name,
            name
        });
    }

    const contextMenu = ContextMenu.create();
    Store.create({ codeBlocks, contextMenu });
    expect(contextMenu.sortedCodeBlocks).toMatchSnapshot();
});