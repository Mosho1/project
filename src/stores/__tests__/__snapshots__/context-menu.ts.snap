// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortedCodeBlocks 1`] = `
Array [
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "ac",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "ac",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "asef",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "asef",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "ba",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "ba",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "ce3",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "ce3",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "cef",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "cef",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "d",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "d",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "dd",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "dd",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "esv",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "esv",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
  Object {
    "code": "(a, b) => a + b",
    "execInputs": Array [],
    "execOutputs": Array [],
    "id": "zzz",
    "inputs": Array [
      Object {
        "defaultValue": null,
        "id": "test/a",
        "name": "a",
        "type": "number",
      },
      Object {
        "defaultValue": null,
        "id": "test/b",
        "name": "b",
        "type": "number",
      },
    ],
    "name": "zzz",
    "returns": Object {
      "defaultValue": null,
      "id": "test/returns",
      "name": "",
      "type": "number",
    },
    "runOnStart": false,
    "values": Array [],
  },
]
`;
